### PR TITLE
Constraints - 1 

### PR DESCRIPTION
**Pull request by JimClarke5**
Adding Constraints.
Constraints impose constraints on weight values.

This does not include the RadialConstraint as that requires `while` operations.

This should be independent class that do not have dependencies on other framework packages.

### CODE REVIEW

File: tensorflow-framework/src/test/java/org/tensorflow/framework/constraints/NonNegTest.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    float[] testValues = {0.1f, 0.5f, 3f, 8f, 1e-7f};
+    for (TestSession.Mode tfMode : tfModes)
+      try (TestSession session = TestSession.createTestSession(tfMode)) {
+        Ops tf = session.getTF();
+        final float[] array = getSampleArray();
+        Operand<TFloat32> weights = tf.reshape(tf.constant(array), tf.constant(Shape.of(100, 100)));
+        for (AtomicInteger i = new AtomicInteger();
+            i.get() < testValues.length;
+            i.getAndIncrement()) {
+          MinMaxNorm<TFloat32> instance =
```
Comment (deansher) [inline tensorflow-framework/src/test/java/org/tensorflow/framework/constraints/NonNegTest.java:52]: From this point forward, the test is erroneously a copy of `MinMaxNormTest`.

File: tensorflow-framework/src/test/java/org/tensorflow/framework/constraints/UnitNormTest.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    float[] testValues = {0.1f, 0.5f, 3f, 8f, 1e-7f};
+    for (TestSession.Mode tfMode : tfModes)
+      try (TestSession session = TestSession.createTestSession(tfMode)) {
+        Ops tf = session.getTF();
+        final float[] array = getSampleArray();
+        Operand<TFloat32> weights = tf.reshape(tf.constant(array), tf.constant(Shape.of(100, 100)));
+        for (AtomicInteger i = new AtomicInteger();
+            i.get() < testValues.length;
+            i.getAndIncrement()) {
+          MaxNorm<TFloat32> instance = new MaxNorm<>(tf, testValues[i.get()]);
```
Comment (deansher) [inline tensorflow-framework/src/test/java/org/tensorflow/framework/constraints/UnitNormTest.java:39]: Erroneously a copy of `MaxNormTest`.

File: tensorflow-framework/src/test/java/org/tensorflow/framework/constraints/UnitNormTest.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    Random rand = new Random(3537L);
+    float[] result = new float[100 * 100];
+    for (int i = 0; i < result.length; i++) {
+      result[i] = rand.nextFloat() * 100 - 50;
+    }
+    result[0] = 0;
+    return result;
+  }
+
+  /** Test of call method, of class MaxNorm. */
```
Comment (deansher) [inline tensorflow-framework/src/test/java/org/tensorflow/framework/constraints/UnitNormTest.java:27]: `MaxNorm` -> `UnitNorm` (not just in the comment, but in the code)

**Review by deansher — CHANGES_REQUESTED** [review]

File: tensorflow-framework/src/main/java/org/tensorflow/framework/constraints/Constraint.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+import org.tensorflow.types.family.TNumber;
+
+import static org.tensorflow.framework.utils.CastHelper.cast;
+
+/**
+ * Base class for Constraints. Constraint subclasses impose constraints on weight values
+ *
+ * @param <T> the date type for the weights
+ */
+public abstract class Constraint<T extends TNumber> {
```
[the remark this answers is not in the document]
Comment (JimClarke5) [reply tensorflow-framework/src/main/java/org/tensorflow/framework/constraints/Constraint.java:28]: No, I will fix.

File: tensorflow-framework/src/main/java/org/tensorflow/framework/constraints/Constraint.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+  }
+
+  /**
+   * Gets the element-wise square root.
+   *
+   * @param x the input Operand.
+   * @return the element-wise square root.
+   */
+  protected Operand<T> sqrt(Operand<T> x) {
+    Class<T> type = x.type();
```
[the remark this answers is not in the document]
Comment (JimClarke5) [reply tensorflow-framework/src/main/java/org/tensorflow/framework/constraints/Constraint.java:67]: Fixed

File: tensorflow-framework/src/main/java/org/tensorflow/framework/constraints/MinMaxNorm.java [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+   * enforcement of the constraint, while rate<1.0 means that weights will be rescaled at each step
+   * to slowly move towards a value inside the desired interval.
+   */
+  private final double rate;
+
+  /** axis along which to calculate weight norms. */
+  private final int[] axes;
+
+  /**
+   * Create a MaxNorm constraint using {@link #MIN_VALUE_DEFAULT} for the min value, {@link
```
[the remark this answers is not in the document]
Comment (JimClarke5) [reply tensorflow-framework/src/main/java/org/tensorflow/framework/constraints/MinMaxNorm.java:52]: Fixed

File: tensorflow-framework/src/main/java/org/tensorflow/framework/constraints/UnitNorm.java [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+  @Override
+  public Operand<T> call(Operand<T> weights) {
+    Class<T> type = weights.type();
+
+    Ops tf = getTF();
+    return tf.math.div(
+        weights,
+        tf.math.add(
+            cast(tf, tf.constant(EPSILON), type),
+            sqrt(
```
[the remark this answers is not in the document]
Comment (JimClarke5) [reply tensorflow-framework/src/main/java/org/tensorflow/framework/constraints/UnitNorm.java:75]: OK

File: tensorflow-framework/src/main/java/org/tensorflow/framework/constraints/UnitNorm.java [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+                tf.reduceSum(
+                    tf.math.square(weights),
+                    tf.constant(getAxes()),
+                    ReduceSum.keepDims(Boolean.TRUE)))));
+  }
+
+  /**
```
[the remark this answers is not in the document]
Comment (JimClarke5) [reply tensorflow-framework/src/main/java/org/tensorflow/framework/constraints/UnitNorm.java:82]: Right now, there are only 3 Constraint classes that use it. I don't think another class is warranted.

File: tensorflow-framework/src/test/java/org/tensorflow/framework/constraints/NonNegTest.java (hunk 1 above)
Comment (JimClarke5) [reply tensorflow-framework/src/test/java/org/tensorflow/framework/constraints/NonNegTest.java:52]: My mistake, I'll fix.

File: tensorflow-framework/src/test/java/org/tensorflow/framework/constraints/UnitNormTest.java (hunk 2 above)
Comment (JimClarke5) [reply tensorflow-framework/src/test/java/org/tensorflow/framework/constraints/UnitNormTest.java:39]: My mistake, I'll fix.

File: tensorflow-framework/src/test/java/org/tensorflow/framework/constraints/UnitNormTest.java (hunk 3 above)
Comment (JimClarke5) [reply tensorflow-framework/src/test/java/org/tensorflow/framework/constraints/UnitNormTest.java:27]: My mistake, I'll fix.

File: tensorflow-framework/src/main/java/org/tensorflow/framework/constraints/Constraint.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (deansher) [reply tensorflow-framework/src/main/java/org/tensorflow/framework/constraints/Constraint.java:67]: 👍 done

File: tensorflow-framework/src/main/java/org/tensorflow/framework/constraints/UnitNorm.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (deansher) [reply tensorflow-framework/src/main/java/org/tensorflow/framework/constraints/UnitNorm.java:70]: 👍

File: tensorflow-framework/src/main/java/org/tensorflow/framework/constraints/Constraint.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (deansher) [reply tensorflow-framework/src/main/java/org/tensorflow/framework/constraints/Constraint.java]: 👍 done

File: tensorflow-framework/src/main/java/org/tensorflow/framework/constraints/Constraint.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (deansher) [reply tensorflow-framework/src/main/java/org/tensorflow/framework/constraints/Constraint.java:67]: 👍 done

File: tensorflow-framework/src/main/java/org/tensorflow/framework/constraints/MinMaxNorm.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (deansher) [reply tensorflow-framework/src/main/java/org/tensorflow/framework/constraints/MinMaxNorm.java]: 👍 done

File: tensorflow-framework/src/main/java/org/tensorflow/framework/constraints/UnitNorm.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (deansher) [reply tensorflow-framework/src/main/java/org/tensorflow/framework/constraints/UnitNorm.java]: 👍 done

File: tensorflow-framework/src/main/java/org/tensorflow/framework/constraints/UnitNorm.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (deansher) [reply tensorflow-framework/src/main/java/org/tensorflow/framework/constraints/UnitNorm.java:70]: 👍

File: tensorflow-framework/src/test/java/org/tensorflow/framework/constraints/NonNegTest.java (hunk 1 above)
Comment (deansher) [reply tensorflow-framework/src/test/java/org/tensorflow/framework/constraints/NonNegTest.java]: 👍 done

File: tensorflow-framework/src/test/java/org/tensorflow/framework/constraints/UnitNormTest.java (hunk 2 above)
Comment (deansher) [reply tensorflow-framework/src/test/java/org/tensorflow/framework/constraints/UnitNormTest.java]: 👍 done

File: tensorflow-framework/src/test/java/org/tensorflow/framework/constraints/UnitNormTest.java (hunk 3 above)
Comment (deansher) [reply tensorflow-framework/src/test/java/org/tensorflow/framework/constraints/UnitNormTest.java]: 👍 done

**Review by deansher — APPROVED** [review]

**Review by karllessard — APPROVED** [review]
All right this one is ready to to, thanks @JimClarke5 and thanks @deansher for reviewing it!

**Pull request merged**